framework/src/auxkernels/ProjectedStatefulMaterialAux.C

```diff
@@ -0,0 +1,75 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ProjectedStatefulMaterialAux.h"
+#include "NodalPatchRecoveryBase.h"
+#include "SerialAccess.h"
+#include "MaterialBase.h"
+#include "Assembly.h"
+
+registerMooseObject("MooseApp", ProjectedStatefulMaterialRealAux);
+registerMooseObject("MooseApp", ADProjectedStatefulMaterialRealAux);
+registerMooseObject("MooseApp", ProjectedStatefulMaterialRealVectorValueAux);
+registerMooseObject("MooseApp", ADProjectedStatefulMaterialRealVectorValueAux);
+registerMooseObject("MooseApp", ProjectedStatefulMaterialRankTwoTensorAux);
+registerMooseObject("MooseApp", ADProjectedStatefulMaterialRankTwoTensorAux);
+registerMooseObject("MooseApp", ProjectedStatefulMaterialRankFourTensorAux);
+registerMooseObject("MooseApp", ADProjectedStatefulMaterialRankFourTensorAux);
+
+template <typename T, bool is_ad>
+InputParameters
+ProjectedStatefulMaterialAuxTempl<T, is_ad>::validParams()
+{
+  auto params = AuxKernel::validParams();
+  params.addClassDescription(
+      "Picks a component of an indexable material property to get projected on an elemental "
+      "Auxvariable. For use by ProjectedStatefulMaterialStorageAction.");
+  params.addRequiredParam<MaterialPropertyName>(
+      "prop", "material property to project onto an elemental aux variable");
+  params.addRequiredParam<unsigned int>("component", "scalar component of the property to fetch");
+  return params;
+}
+
+template <typename T, bool is_ad>
+ProjectedStatefulMaterialAuxTempl<T, is_ad>::ProjectedStatefulMaterialAuxTempl(
+    const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _current_subdomain_id(_assembly.currentSubdomainID()),
+    _prop(getGenericMaterialProperty<T, is_ad>("prop")),
+    _component(getParam<unsigned int>("component"))
+{
+}
+
+template <typename T, bool is_ad>
+void
+ProjectedStatefulMaterialAuxTempl<T, is_ad>::initialSetup()
+{
+  // get all material classes that provide properties for this object
+  _required_materials = buildRequiredMaterials();
+}
+
+template <typename T, bool is_ad>
+Real
+ProjectedStatefulMaterialAuxTempl<T, is_ad>::computeValue()
+{
+  if (_t_step == 0)
+    for (const auto & mat : _required_materials[_current_subdomain_id])
+      mat->initStatefulProperties(_qrule->size());
```

Comment:
If I look at the execution order in `FEProblemBase::initialSetup` I believe we should be initializing stateful material properties before we call `execute(EXEC_INITIAL)` ... so I think this should be unnecessary

Reply:
Yeah, but we then move those values into the old state and immediately overwrite the material values with a new evaluateion from computeQpProperties. Plus, if we do not make the material stateful MOOSE will never call initStatefulProperties for us.

Reply:
Man this PR is a definite thinker...

> Yeah, but we then move those values into the old state and immediately overwrite the material values with a new evaluateion from computeQpProperties.

I'm not quite sure what by "move". In your new system, what are you moving? In the traditional stateful material property system we do not move, but copy the values from the current props populated by `initStatefulProps` into the old and older states during stateful material property initialization. (We then swap/move when advancing state at the end of time steps).

I have realized now that you are spot-on about a later `computeQpProperties` evaluation (during `_t_step == 0`) inaccurately writing the current property because in the case of stateful material properties your current material property evaluation depends on old/older material property information..... Which means that *any* auxiliary kernel executed on initial which uses a stateful material property and does not do what you are doing here will get a wrong result!? That is what I've discovered when doing some slight modifications to the `test/tests/materials/stateful_coupling/stateful_aux.i`.

Reply:
I really don't like this. This is at the middle of the confusion I have to go through when working on stateful + restart. 

I dont like the idea of having any non-system call through to that init call. There's so many opportunities for error here

Reply:
The problem is that without this, the result is wrong. And it's wrong for any UO or Aux that doesn't do this when using a stateful material property. So for instance, `MaterialRealAux` is wrong on initial if it's acting with a stateful material property

Reply:
There are no old properties in my system

Reply:
If you imagine the exact same code as above but imagine that `_properties_old` and `_properties_older` were instead generated by your `InterpolatedStatefulMaterial` class, then don't you have the exact same problem after a call to `StatefulTest::computeQpProperties`? Let's imagine two scenarios:

1. Your aux kernel somehow executes before a call to `reinitMaterials`. What will the result of a call to `StatefulTest::computeQpProperties` be for `_properties[qp]`? It will be 4, right? We want objects reading this property to get 2.
2. Your aux kernel does not execute before a call to `reinitMaterials`. What will the result of a call to `computeQpProperties` be for `_properties[qp]`? It will be two times the initial condition of your aux variable, which is most likely 0. We want objects reading this property to get 2

Reply:
I don't think this is a significant issue and I don't think you could have done any better within the constraints of our order of operations. I'm just saying there's a corner case if someone somehow decided to rely on stateful material property values read during `EXEC_INITIAL` (post a `computeQpProperties` call)

Reply:
I've opened https://github.com/idaholab/moose/issues/26281 for this